In [1]:
import arviz as az
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import pymc as pm

%load_ext watermark
%load_ext lab_black

# Taste of cheese

Adapted from [Unit 6: cheese.odc](https://raw.githubusercontent.com/areding/6420-pymc/main/original_examples/Codes4Unit6/cheese.odc).

The link in the original .odc file is dead. I downloaded the data from [here](https://www3.nd.edu/~busiforc/handouts/Data%20and%20Stories/multicollinearity/Cheese%20Taste/Cheddar%20Cheese%20Data.html) and have a copy [here](https://raw.githubusercontent.com/areding/6420-pymc/main/data/cheese.csv).

Associated lecture videos 
Unit 6 lesson 5

Unit 6 Lesson 8

## Problem statement

As cheddar cheese matures, a variety of chemical processes take place. The taste of matured cheese is related to the concentration of several chemicals in the final product. In a study of cheddar cheese from the LaTrobe Valley of Victoria, Australia, samples of cheese were analyzed for their chemical composition and were subjected to taste tests. Overall taste scores were obtained by combining the scores from several tasters.

Can the score be predicted well by the predictors: Acetic, H2S, and Lactic?

In [2]:
data = pd.read_csv("../data/cheese.csv", index_col=0)
X = data[["Acetic", "H2S", "Lactic"]].to_numpy()
# add intercept column to X
X_aug = np.concatenate((np.ones((X.shape[0], 1)), X), axis=1)
y = data["taste"].to_numpy()

In [3]:
with pm.Model() as m:
    # associate data with model (this makes prediction easier)
    X_data = pm.Data("X", X_aug, mutable=True)
    y_data = pm.Data("y", y, mutable=False)

    # priors
    beta = pm.Normal("beta", mu=0, sigma=1000, shape=X.shape[1] + 1)
    tau = pm.Gamma("tau", alpha=0.001, beta=0.001)
    sigma = 1 / pm.math.sqrt(tau)

    mu = pm.math.dot(X_data, beta)

    # likelihood
    taste_score = pm.Normal("taste_score", mu=mu, sigma=sigma, observed=y_data)

    # start sampling
    trace = pm.sample(
        5000,  # samples
        chains=4,
        tune=1000,
        init="jitter+adapt_diag",
        random_seed=1,
        cores=4,
        return_inferencedata=True,
        target_accept=0.95,
    )

Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, tau]


Sampling 4 chains for 1_000 tune and 5_000 draw iterations (4_000 + 20_000 draws total) took 26 seconds.


In [4]:
az.summary(trace, hdi_prob=0.95)

mean      sd  hdi_2.5%  hdi_97.5%  mcse_mean  mcse_sd  ess_bulk  \
beta[0] -28.992  20.360   -70.973      9.987      0.246    0.174    6846.0   
beta[1]   0.368   4.604    -8.545      9.665      0.058    0.041    6363.0   
beta[2]   3.912   1.303     1.265      6.402      0.014    0.010    9069.0   
beta[3]  19.589   8.935     1.944     37.349      0.089    0.063   10099.0   
tau       0.010   0.003     0.005      0.015      0.000    0.000   10787.0   

         ess_tail  r_hat  
beta[0]    9625.0    1.0  
beta[1]    8688.0    1.0  
beta[2]   10989.0    1.0  
beta[3]   10269.0    1.0  
tau       11989.0    1.0

Results are pretty close to OpenBUGS:

|         mean |     sd   |        MC_error |   val2.5pc | median  |   val97.5pc | start   | sample |        |
|--------------|----------|-----------------|------------|---------|-------------|---------|--------|--------|
| beta0        | -29.75   | 20.24           | 0.7889     | -70.06  | -29.75      | 11.11   | 1000   | 100001 |
| beta1        | 0.4576   | 4.6             | 0.189      | -8.716  | 0.4388      | 9.786   | 1000   | 100001 |
| beta2        | 3.906    | 1.291           | 0.02725    | 1.345   | 3.912       | 6.47    | 1000   | 100001 |
| beta3        | 19.79    | 8.893           | 0.2379     | 2.053   | 19.88       | 37.2    | 1000   | 100001 |
| tau          | 0.009777 | 0.002706        | 2.29E-05   | 0.00522 | 0.009528    | 0.01575 | 1000   | 100001 |

PyMC gives some warnings about the model unless we increase the ```target_accept``` parameter of ```pm.sample```, while BUGS doesn't. This is because PyMC uses more diagnostics to check if there are any problems with its exploration of the parameter space. Divergences indicate bias in the results. BUGS will happily run this model without reporting any problems, but it doesn't mean that there aren't any.

For further reading, check out [Diagnosing Biased Inference with Divergences](https://docs.pymc.io/en/v3/pymc-examples/examples/diagnostics_and_criticism/Diagnosing_biased_Inference_with_Divergences.html).

It looks like there are multiple ways to get predictions on out-of-sample data in PyMC. The easiest way is to set up a shared variable using pm.Data in the original model, then using pm.set_data to change to the new observations before calling pm.sample_posterior_predictive.

In [5]:
# single prediction on out-of-sample data
new_obs = np.array([[1.0, 5.0, 7.1, 1.5]])
pm.set_data({"X": new_obs}, model=m)
ppc = pm.sample_posterior_predictive(trace, model=m, predictions=True)

Sampling: [taste_score]


The default behavior now is to give one prediction per y-value. The professor often asks for a single prediction based on the new data; the equivalent here would be to take the mean of the predicted values.

In [6]:
ppc

Inference data with groups:
	> predictions
	> predictions_constant_data

In [7]:
az.summary(ppc.predictions, hdi_prob=0.95, kind="stats")

mean      sd  hdi_2.5%  hdi_97.5%
taste_score[0]   30.034  11.206     8.428     52.410
taste_score[1]   30.024  11.240     7.574     52.214
taste_score[2]   30.042  11.171     7.675     51.805
taste_score[3]   29.951  11.109     7.596     51.404
taste_score[4]   29.965  11.292     7.669     52.137
taste_score[5]   29.855  11.226     7.983     52.260
taste_score[6]   30.016  11.297     7.027     51.668
taste_score[7]   29.950  11.197     7.655     51.733
taste_score[8]   30.088  11.153     7.976     51.847
taste_score[9]   30.013  11.305     7.807     52.522
taste_score[10]  30.074  11.280     7.991     52.251
taste_score[11]  30.077  11.301     8.063     52.666
taste_score[12]  30.018  11.153     8.358     51.977
taste_score[13]  29.907  11.199     8.233     52.328
taste_score[14]  30.038  11.224     7.880     51.973
taste_score[15]  30.059  11.219     7.220     51.352
taste_score[16]  30.046  11.152     8.293     52.130
taste_score[17]  30.083  11.278     7.887     52.160
taste_score[18]  29.892  11.129     7.790     51.584
taste_score[19]  30.186  11.094     7.468     51.241
taste_score[20]  30.010  11.189     8.172     52.384
taste_score[21]  30.219  11.188     8.085     52.284
taste_score[22]  29.952  11.177     8.044     52.130
taste_score[23]  29.923  11.198     7.129     51.547
taste_score[24]  29.912  11.194     8.495     52.449
taste_score[25]  30.063  11.106     7.633     51.660
taste_score[26]  30.049  11.158     7.769     51.751
taste_score[27]  29.867  11.220     7.150     51.627
taste_score[28]  30.062  11.078     7.869     51.349
taste_score[29]  29.921  11.086     8.184     52.025

In [8]:
ppc.predictions

<xarray.Dataset>
Dimensions:            (chain: 4, draw: 5000, taste_score_dim_0: 30)
Coordinates:
  * chain              (chain) int64 0 1 2 3
  * draw               (draw) int64 0 1 2 3 4 5 ... 4995 4996 4997 4998 4999
  * taste_score_dim_0  (taste_score_dim_0) int64 0 1 2 3 4 5 ... 25 26 27 28 29
Data variables:
    taste_score        (chain, draw, taste_score_dim_0) float64 43.78 ... 7.525
Attributes:
    created_at:                 2022-11-03T01:04:14.100924
    arviz_version:              0.13.0
    inference_library:          pymc
    inference_library_version:  4.2.2

In [9]:
az.summary(ppc.predictions.mean(dim=("taste_score_dim_0")), hdi_prob=0.95)

mean     sd  hdi_2.5%  hdi_97.5%  mcse_mean  mcse_sd  ess_bulk  \
taste_score  30.01  4.207     21.46     38.061      0.044    0.031    9319.0   

             ess_tail  r_hat  
taste_score   12692.0    1.0

In [10]:
%watermark -n -u -v -iv -p aesara,aeppl

Last updated: Thu Nov 03 2022

Python implementation: CPython
Python version       : 3.10.6
IPython version      : 8.5.0

aesara: 2.8.7
aeppl : 0.0.38

pandas    : 1.5.1
matplotlib: 3.6.1
arviz     : 0.13.0
numpy     : 1.23.4
pymc      : 4.2.2

